server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileSearchAction.java

```diff
@@ -0,0 +1,171 @@
+import org.sonar.api.resources.Language;
+import org.sonar.api.resources.Languages;
+import org.sonar.api.server.ws.*;
+import org.sonar.api.server.ws.WebService.NewAction;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.core.qualityprofile.db.QualityProfileDto;
+import org.sonar.server.qualityprofile.QProfile;
+import org.sonar.server.qualityprofile.QProfileLookup;
+import org.sonar.server.qualityprofile.QProfileService;
+
+import java.util.*;
+
+public class QProfileSearchAction implements RequestHandler {
+
+  private static final String FIELD_KEY = "key";
+  private static final String FIELD_NAME = "name";
+  private static final String FIELD_LANGUAGE = "language";
+  private static final String FIELD_IS_INHERITED = "isInherited";
+  private static final String FIELD_IS_DEFAULT = "isDefault";
+  private static final String FIELD_PARENT_KEY = "parentKey";
+  private static final Set<String> ALL_FIELDS = ImmutableSet.of(FIELD_KEY, FIELD_NAME, FIELD_LANGUAGE, FIELD_IS_INHERITED, FIELD_PARENT_KEY, FIELD_IS_DEFAULT);
+
+  private static final String PARAM_LANGUAGE = FIELD_LANGUAGE;
+  private static final String PARAM_FIELDS = "f";
+
+
+  private final Languages languages;
+
+  private final QProfileLookup profileLookup;
+
+  private final QProfileService profileService;
+
+  public QProfileSearchAction(Languages languages, QProfileLookup profileLookup, QProfileService profileService) {
+    this.languages = languages;
+    this.profileLookup = profileLookup;
+    this.profileService = profileService;
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    List<String> fields = request.paramAsStrings(PARAM_FIELDS);
+
+    String language = request.param(PARAM_LANGUAGE);
+
+    List<QProfile> profiles = null;
+    Map<String, String> defaultProfileKeyByLanguage = Maps.newHashMap();
+    if (language == null) {
+      profiles = profileLookup.allProfiles();
+      for (Language lang : languages.all()) {
+        String langKey = lang.getKey();
+        cacheDefaultProfileIfExists(langKey, defaultProfileKeyByLanguage);
+      }
+    } else {
+      profiles = profileLookup.profiles(language);
+      cacheDefaultProfileIfExists(language, defaultProfileKeyByLanguage);
+    }
+
+    Collections.sort(profiles, new Comparator<QProfile>() {
+      @Override
+      public int compare(QProfile o1, QProfile o2) {
+        int compare = o1.language().compareTo(o2.language());
+        if (compare == 0) {
+          compare = o1.name().compareTo(o2.name());
+        }
+        return compare;
+      }
+    });
+
+    JsonWriter json = response.newJsonWriter().beginObject();
+    writeProfiles(json, profiles, defaultProfileKeyByLanguage, fields);
+    writeLanguages(json);
+    json.endObject().close();
+  }
+
+  private void cacheDefaultProfileIfExists(String langKey, Map<String, String> defaultByLanguage) {
+    QualityProfileDto defaultProfile = profileService.getDefault(langKey);
```

Comment:
If there's 20 languages installed, 20 db queries will be done. I think we should find a way to easily query for all default profiles in once, no ? :)